File: typescript/graph/bfs.ts
import Graph from "./graph.ts";
import Queue from "./queue.ts";

const vertexColor = ["White", "Gray", "Black"] as const;
type VertexColor = typeof vertexColor[number];

const bfsSearch = (
  graph: Graph,
  start: number
): { dist: number[]; pred: number[] } => {
  const n = graph.getVerticesNum();
  const pred: number[] = Array(n).fill(-1);
  const dist: number[] = Array(n).fill(Number.MAX_SAFE_INTEGER);
  const color: VertexColor[] = Array(n).fill("White");

  dist[start] = 0;
  color[start] = "Gray";

  const q = new Queue<number>();
  q.enqueue(start);

  while (!q.isEmpty()) {
    const u = q.peek() as number;

    for (const [v, w] of graph.getVertices(u)) {
      if (color[v] === "White") {
        dist[v] = dist[u] + 1;
        pred[v] = u;
        color[v] = "Gray";
        q.enqueue(v);
      }
    }

    q.dequeue();
    color[u] = "Black";
  }

  return { pred, dist };
};

export default bfsSearch;

File: python/graph/bfs.py
import sys
from enum import Enum
from queue import Queue
from .graph import Graph


class Color(Enum):
    White = "White"
    Gray = "Gray"
    Black = "Black"


def bfs_search(graph: Graph, start: int) -> list[int]:
    n: int = graph.get_vertices_num()
    pred: list[int] = [-1 for _ in range(n)]
    dist: list[int] = [sys.maxsize for _ in range(n)]
    colors: list[Color] = [Color.White for _ in range(n)]

    dist[start] = 0
    colors[start] = Color.Gray

    q: Queue = Queue()
    q.put(start)

    while not q.empty():
        u: int = q.get()
        for (v, w) in graph.get_vertices(u):
            if colors[v] == Color.White:
                dist[v] = dist[u] + 1
                pred[v] = u
                colors[v] = Color.Gray
                q.put(v)

        q.get()
        colors[u] = Color.Black

    return pred
